website/blog/2023-02-21.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,530 @@
+---
+title: "How MongoDB compares arrays in filtering"
+slug: ferretdb-compare-arrays
+authors: [chi]
+description: In this article, we explore in detail how MongoDB BSON arrays are compared
+in filter operators.
+keywords:
+  - MongoDB sorting
+  - MongoDB array comparison
+image: /img/blog/ferretdb-how-mongodb-compares-arrays.jpg
+unlisted: true
+---
+
+In this article, we explore in detail how MongoDB BSON arrays are compared in filter operators.
+
+![How MongoDB compares arrays](/img/blog/ferretdb-how-mongodb-compares-arrays.jpg)
+
+<!--truncate-->
+
+## Understanding BSON Comparison for Filtering
+
+The purpose of comparing BSON values is to determine which value is
+equal, greater or less than the other.
+The comparison is used in various operations, such as sorting, filter operators,
+as well as update operators.
+When comparing values of different BSON types, the [BSON Comparison Order](#bson-comparison-order)
+is used to determine the result.
+However, in the case of array comparison against other BSON value,
+the comparison may use the BSON comparison order or an element from the array.
+
+In this blog post, we will discuss the comparison of BSON values used in filter operators,
+with a focus on arrays.
+We will provide examples to help clarify the process.
+
+### BSON Comparison Order
+
+BSON values have a defined order of comparison, listed from lowest to highest as follows:
+
+1. Null
+2. Numbers
+3. String
+4. Object
+5. Array
+6. BinData
+7. ObjectId
+8. Boolean
+9. Date
+10. Timestamp
+11. Regular Expression
+
+The BSON comparison order is used for comparing different BSON types.
+A BSON array
+may use an element from an array to compare with other BSON type.
+We will discuss in detail in the following section.
+
+All number types are the same BSON comparison order, regardless of
+`double`, `int32`, `int64` or `decimal`.
+
+### Comparison Rules of Filter
+
+When comparing an array filter argument with array collections,
```

Comment:
I guess "array collections" are data stored in the DB, not collections that contain arrays, right?
Do we need to add some terminology - like "we will call XYZ filter arguments and XYZ array collections"?

Reply:
I'll add a terminology _array values_